File: packages/aws-cdk-dotnet/src/Amazon.CDK/PolicyDocument.cs
using AWS.Jsii.Runtime.Deputy;

namespace Amazon.CDK
{
    [JsiiClass(typeof(PolicyDocument), "@aws-cdk/cdk.PolicyDocument", "[{\"name\":\"baseDocument\",\"type\":{\"primitive\":\"any\",\"optional\":true}}]")]
    public class PolicyDocument : Token
    {
        public PolicyDocument(object baseDocument): base(new DeputyProps(new object[]{baseDocument}))
        {
        }

        protected PolicyDocument(ByRefValue reference): base(reference)
        {
        }

        protected PolicyDocument(DeputyProps props): base(props)
        {
        }

        [JsiiProperty("isEmpty", "{\"primitive\":\"boolean\"}")]
        public virtual bool IsEmpty
        {
            get => GetInstanceProperty<bool>();
        }

        /// <summary>
        /// The number of statements already added to this policy.
        /// Can be used, for example, to generate uniuqe "sid"s within the policy.
        /// </summary>
        [JsiiProperty("statementCount", "{\"primitive\":\"number\"}")]
        public virtual double StatementCount
        {
            get => GetInstanceProperty<double>();
        }

        [JsiiMethod("resolve", "{\"primitive\":\"any\"}", "[]")]
        public override object Resolve()
        {
            return InvokeInstanceMethod<object>(new object[]{});
        }

        [JsiiMethod("addStatement", "{\"fqn\":\"@aws-cdk/cdk.PolicyDocument\"}", "[{\"name\":\"statement\",\"type\":{\"fqn\":\"@aws-cdk/cdk.PolicyStatement\"}}]")]
        public virtual PolicyDocument AddStatement(PolicyStatement statement)
        {
            return InvokeInstanceMethod<PolicyDocument>(new object[]{statement});
        }
    }
}

File: packages/@aws-cdk/aws-iot/lib/policy-document.ts
import * as cdk from '@aws-cdk/core';
import { PolicyStatement } from './policy-statement';

/**
 * Properties for a new PolicyDocument
 */
export interface PolicyDocumentProps {
  /**
   * Initial statements to add to the policy document
   *
   * @default - No statements
   */
  readonly statements?: PolicyStatement[];
}

/**
 * A PolicyDocument is a collection of statements
 */
export class PolicyDocument implements cdk.IResolvable {

  /**
   * Creates a new PolicyDocument based on the object provided.
   * This will accept an object created from the `.toJSON()` call
   * @param obj the PolicyDocument in object form.
   */
  public static fromJson(obj: any): PolicyDocument {
    const newPolicyDocument = new PolicyDocument();
    const statement = obj.Statement ?? [];
    if (statement && !Array.isArray(statement)) {
      throw new Error('Statement must be an array');
    }
    newPolicyDocument.addStatements(...obj.Statement.map((s: any) => PolicyStatement.fromJson(s)));
    return newPolicyDocument;
  }

  public readonly creationStack: string[];
  private readonly statements = new Array<PolicyStatement>();

  constructor(props: PolicyDocumentProps = {}) {
    this.creationStack = cdk.captureStackTrace();

    this.addStatements(...props.statements || []);
  }

  public resolve(context: cdk.IResolveContext): any {
    context.registerPostProcessor(new RemoveDuplicateStatements());
    return this.render();
  }

  /**
   * Whether the policy document contains any statements.
   */
  public get isEmpty(): boolean {
    return this.statements.length === 0;
  }

  /**
   * The number of statements already added to this policy.
   */
  public get statementCount(): number {
    return this.statements.length;
  }

  /**
   * Adds a statement to the policy document.
   *
   * @param statement the statement to add.
   */
  public addStatements(...statement: PolicyStatement[]) {
    this.statements.push(...statement);
  }

  /**
   * Encode the policy document as a string
   */
  public toString() {
    return cdk.Token.asString(this, {
      displayHint: 'PolicyDocument',
    });
  }

  /**
   * JSON-ify the document
   *
   * Used when JSON.stringify() is called
   */
  public toJSON() {
    return this.render();
  }

  /**
   * Validate that all policy statements in the policy document satisfies the
   * requirements for any policy.
   *
   * @see https://docs.aws.amazon.com/iot/latest/developerguide/iot-policies.html
   */
  public validateForAnyPolicy(): string[] {
    const errors = new Array<string>();
    for (const statement of this.statements) {
      errors.push(...statement.validateForAnyPolicy());
    }
    return errors;
  }

  private render(): any {
    if (this.isEmpty) {
      return undefined;
    }

    const doc = {
      Statement: this.statements.map(s => s.toStatementJson()),
      Version: '2012-10-17',
    };

    return doc;
  }
}

/**
 * Removes duplicate statements
 */
class RemoveDuplicateStatements implements cdk.IPostProcessor {
  constructor() {
  }

  public postProcess(input: any, _context: cdk.IResolveContext): any {
    if (!input || !input.Statement) {
      return input;
    }

    const jsonStatements = new Set<string>();
    const statements: any[] = [];

    for (const statement of input.Statement) {
      const jsonStatement = JSON.stringify(statement);
      if (!jsonStatements.has(jsonStatement)) {
        statements.push(statement);
        jsonStatements.add(jsonStatement);
      }
    }

    return {
      ...input,
      Statement: statements,
    };
  }
}
